/*
*  Copyright (c) 2020, The OpenThread Authors.
*  Copyright 2023-2024 NXP
*  All rights reserved.
*
*  Redistribution and use in source and binary forms, with or without
*  modification, are permitted provided that the following conditions are met:
*  1. Redistributions of source code must retain the above copyright
*     notice, this list of conditions and the following disclaimer.
*  2. Redistributions in binary form must reproduce the above copyright
*     notice, this list of conditions and the following disclaimer in the
*     documentation and/or other materials provided with the distribution.
*  3. Neither the name of the copyright holder nor the
*     names of its contributors may be used to endorse or promote products
*     derived from this software without specific prior written permission.
*
*  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
*  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
*  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
*  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
*  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
*  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
*  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
*  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
*  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
*  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
*  POSSIBILITY OF SUCH DAMAGE.
*/

/*
* Header for patch information.
* Description           : Retry mechanism if mTxRadioTid is expected and current command failed
*                         with OT_ERROR_BUSY.
* Version               : v002
* Mandatory/Optional    : Mandatory
* Impact if not applied : Under corner scenario, OT host SW may crash.
*/

diff --git a/src/lib/spinel/radio_spinel.cpp b/src/lib/spinel/radio_spinel.cpp
index 0347d53e4..43e5624f9 100644
--- a/src/lib/spinel/radio_spinel.cpp
+++ b/src/lib/spinel/radio_spinel.cpp
@@ -1489,6 +1489,19 @@ otError RadioSpinel::RequestV(uint32_t command, spinel_prop_key_t aKey, const ch
     VerifyOrExit(tid > 0, error = OT_ERROR_BUSY);
 
     error = mSpinelDriver.SendCommand(command, aKey, tid, aFormat, aArgs);
+
+    if( mTxRadioTid && (error == OT_ERROR_BUSY) )
+    {
+        // Handle RETRY mechanism in case of VALUE_SET / STREAM_RAW failed transaction
+        error = WaitResponse();
+
+        if( error == OT_ERROR_NONE )
+        {
+            // Call again the current command. SPI interface shall be available now !
+            error = mSpinelDriver.SendCommand(command, aKey, tid, aFormat, aArgs);
+        }
+    }
+
     SuccessOrExit(error);
 
     if (aKey == SPINEL_PROP_STREAM_RAW)
